ci: golangci-lint v2.12.1 detected issues

When running the latest version of `golangci-lint` on `main`, the following issues are detected:

```
$ golangci-lint version
golangci-lint has version 2.12.1 built with go1.26.2 from 9aa24e94 on 2026-05-02T10:12:50Z

$ golangci-lint run ./...
cmd/readiness-condition-reporter/main.go:155:40: G704: SSRF via taint analysis (gosec)
        req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
                                              ^
cmd/readiness-condition-reporter/main.go:164:24: G704: SSRF via taint analysis (gosec)
        resp, err := client.Do(req)
                              ^
internal/controller/suite_test.go:155:23: G703: Path traversal via taint analysis (gosec)
                if _, err := os.Stat(assets); err == nil {
                                    ^
internal/webhook/nodereadinessgaterule_webhook.go:49:6: Consider preallocating allErrs (prealloc)
        var allErrs field.ErrorList
            ^
4 issues:
* gosec: 3
* prealloc: 1
```
We need to either fix those issues or disable those checks, be it via config or directives in the code.

These issues block upgrading to the latest version of `golangci-lint`, since it will break in CI.

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -20,4 +20,4 @@ jobs:
       - name: Run linter
         uses: golangci/golangci-lint-action@4afd733a84b1f43292c63897423277bb7f4313a9 # v8.0.0
         with:
-          version: v2.4.0
+          version: v2.12.1

--- a/Makefile
+++ b/Makefile
@@ -48,7 +48,7 @@ SETUP_ENVTEST_PKG := sigs.k8s.io/controller-runtime/tools/setup-envtest
 KUBEBUILDER_ENVTEST_KUBERNETES_VERSION ?= $(shell go list -m -f "{{ .Version }}" k8s.io/api | awk -F'[v.]' '{printf "1.%d", $$3}')
 
 GOLANGCI_LINT_BIN := golangci-lint
-GOLANGCI_LINT_VER ?= v2.9.0
+GOLANGCI_LINT_VER ?= v2.12.1
 GOLANGCI_LINT := $(abspath $(TOOLS_BIN_DIR)/$(GOLANGCI_LINT_BIN)-$(GOLANGCI_LINT_VER))
 GOLANGCI_LINT_PKG := github.com/golangci/golangci-lint/v2/cmd/golangci-lint
 

--- a/cmd/readiness-condition-reporter/main.go
+++ b/cmd/readiness-condition-reporter/main.go
@@ -21,6 +21,7 @@ import (
 	"fmt"
 	"io"
 	"net/http"
+	"net/url"
 	"os"
 	"os/signal"
 	"syscall"
@@ -58,18 +59,22 @@ func main() {
 	nodeName := os.Getenv(envNodeName)
 	if nodeName == "" {
 		klog.ErrorS(nil, "Environment variable not set", "variable", envNodeName)
+		klog.Flush()
 		os.Exit(1)
 	}
 
 	conditionType := os.Getenv(envConditionType)
 	if conditionType == "" {
 		klog.ErrorS(nil, "Environment variable not set", "variable", envConditionType)
+		klog.Flush()
 		os.Exit(1)
 	}
 
-	checkEndpoint := os.Getenv(envCheckEndpoint)
-	if checkEndpoint == "" {
-		klog.ErrorS(nil, "Environment variable not set", "variable", envCheckEndpoint)
+	checkEndpoint, err := validateCheckEndpoint(os.Getenv(envCheckEndpoint))
+
+	if err != nil {
+		klog.ErrorS(err, "Invalid check endpoint configuration", "variable", envCheckEndpoint)
+		klog.Flush()
 		os.Exit(1)
 	}
 
@@ -90,6 +95,7 @@ func main() {
 	config, err := rest.InClusterConfig()
 	if err != nil {
 		klog.ErrorS(err, "Failed to create in-cluster config")
+		klog.Flush()
 		os.Exit(1)
 	}
 
@@ -104,6 +110,7 @@ func main() {
 	clientset, err := kubernetes.NewForConfig(config)
 	if err != nil {
 		klog.ErrorS(err, "Failed to create client")
+		klog.Flush()
 		os.Exit(1)
 	}
 
@@ -150,9 +157,29 @@ func runCheck(ctx context.Context, httpClient *http.Client, clientset kubernetes
 	}
 }
 
+// validateCheckEndpoint ensures the health check endpoint is a well-formed HTTP(S) URL.
+func validateCheckEndpoint(endpoint string) (string, error) {
+	if endpoint == "" {
+		return "", fmt.Errorf("endpoint cannot be empty")
+	}
+
+	parsed, err := url.Parse(endpoint)
+	if err != nil {
+		return "", fmt.Errorf("invalid endpoint URL: %w", err)
+	}
+	if parsed.Scheme != "http" && parsed.Scheme != "https" {
+		return "", fmt.Errorf("unsupported scheme %q: must be http or https", parsed.Scheme)
+	}
+	if parsed.Host == "" {
+		return "", fmt.Errorf("missing host in endpoint URL")
+	}
+
+	return endpoint, nil
+}
+
 // checkHealth performs an HTTP request to check component health.
 func checkHealth(ctx context.Context, client *http.Client, endpoint string) (*HealthResponse, error) {
-	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil) //nolint:gosec // endpoint validated at startup
 	if err != nil {
 		return &HealthResponse{
 			Healthy: false,
@@ -161,7 +188,7 @@ func checkHealth(ctx context.Context, client *http.Client, endpoint string) (*He
 		}, nil
 	}
 
-	resp, err := client.Do(req)
+	resp, err := client.Do(req) //nolint:gosec // endpoint validated at startup
 	if err != nil {
 		return &HealthResponse{
 			Healthy: false,

--- a/internal/controller/suite_test.go
+++ b/internal/controller/suite_test.go
@@ -152,8 +152,9 @@ var _ = AfterSuite(func() {
 func getFirstFoundEnvTestBinaryDir() string {
 	// First check if KUBEBUILDER_ASSETS is already set
 	if assets := os.Getenv("KUBEBUILDER_ASSETS"); assets != "" {
-		if _, err := os.Stat(assets); err == nil {
-			return assets
+		cleaned := filepath.Clean(assets)
+		if _, err := os.Stat(cleaned); err == nil {
+			return cleaned
 		}
 	}
 

--- a/internal/webhook/nodereadinessgaterule_webhook.go
+++ b/internal/webhook/nodereadinessgaterule_webhook.go
@@ -46,7 +46,7 @@ func NewNodeReadinessRuleWebhook(c client.Client) *NodeReadinessRuleWebhook {
 
 // validateNodeReadinessRule performs validation logic.
 func (w *NodeReadinessRuleWebhook) validateNodeReadinessRule(ctx context.Context, rule *readinessv1alpha1.NodeReadinessRule, isUpdate bool) field.ErrorList {
-	var allErrs field.ErrorList
+	allErrs := make(field.ErrorList, 0, 4)
 
 	// Validate basic fields
 	allErrs = append(allErrs, w.validateSpec(rule.Spec)...)

